File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        unordered_set<int> s;
        int m=grid.size();
        int n=grid[0].size();
        int ans=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]=='0' || s.find(i*n+j)!=s.end())
                    continue;
                dfs(grid,s,i,j,m,n);
                ans++;
            }
        }  
        return ans;  
    }

    void dfs(vector<vector<char>>& grid, unordered_set<int> &s, int i, int j, int m, int n){
        if(i<0||j<0||i>=m||j>=n||grid[i][j]=='0'||s.find(i*n+j)!=s.end())
            return;
        s.insert(i*n+j);
        dfs(grid,s,i+1,j,m,n);
        dfs(grid,s,i,j+1,m,n);
        dfs(grid,s,i-1,j,m,n);
        dfs(grid,s,i,j-1,m,n);
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m=len(grid)
        n=len(grid[0])
        ma=0
        for i in range(m):
            for j in range(n):
                if(grid[i][j]):
                    ma=max(ma,self.dfs(grid,i,j,m,n))
        return ma
    
    def dfs(self,grid,i,j,m,n):
        if i<0 or j<0 or i>=m or j>=n or not grid[i][j]:
            return 0
        grid[i][j]=0
        return 1+self.dfs(grid,i+1,j,m,n) + self.dfs(grid,i-1,j,m,n) + self.dfs(grid,i,j+1,m,n) +self.dfs(grid,i,j-1,m,n)


                
        